### PR TITLE
JBPM-5568 Cannot open Diagrams that contain "unsupported" Definitions

### PR DESCRIPTION
**Pull request by tiagodolphine**
When loading a diagram that contains unsupported definitions on Stunner, than an error message is displayed.It was created `DiagramClientErrorHandler` to handle the error generating a message based on the `ClientTranslationService` using the bundle properties that contains the message template on `StunnerCoreConstants.properties`.

**Note1** : On the standalone showcases, it was added an exception handling for the unsupported diagrams on the `AbstractVFSDiagramService.getDiagramsByPath` not returning the diagram with marshalling error, only the supported ones.
**Note2**: `DiagramServiceImpl` was annotated with `@ApplicationScoped`, on the tests I have noticed that the class was being instantiated by the CDI every time it was injected, it could be a singleton, and since we are registering all diagrams on the `@PostConstruct` this was not a good behavior.

@romartin
@hasys
@manstis

### CODE REVIEW

**Comment by romartin** [comment]
jenkins execute full downstream build

File: kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java [hunk 1]
```diff
@@ -631,7 +637,7 @@ private void onSaveSuccess() {
     }
 
     private void onSaveError(final ClientRuntimeError error) {
-        showError(error.toString());
+        showError(error.getThrowable() instanceof DefinitionNotFoundException ? "somps" : error.getMessage());
```
Comment (manstis) [inline kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java]: "somps"? What's this?

Comment (tiagodolphine) [reply kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java]: Hi @manstis , thanks man I've forgotten this point :( it was just a test.
Already fixed and pushed.

**Comment by hasys** [comment]
Jenkins execute full downstream build

**Comment by romartin** [comment]
please @hasys , need your review for merging it. Thanks!

**Comment by hasys** [comment]
Jenkins execute full downstream build

**Comment by tiagodolphine** [comment]
Hi @hasys, I hope you're better 
About the empty screen, are you referring to standalone or project/wb ? Can you give more details? Because, on project and wb you get an error message if you try to open a diagram with unsupported elements and on standalone the diagram is not listed so you cannot even try to open it.

**Comment by hasys** [comment]
Hi @tiagodolphine 
Yes, I am better now, thank you!

I mean in KIE Workbench:
image

When I click on **OK** button I will see Sunner without canvas (nothing will changed after **OK** click, just notification will be hidden). I see real use case for it (Use **Latest Version** to restore previous stable state), but other buttons produce errors. For example when I click on **Save** button:
image

**Comment by romartin** [comment]
ohh good point @hasys , well seen. 
@tiagodolphine that was not in the initial plan for this task, not sure if it's easy to fix at this point, but for me makes sense what @hasys says - we should a nice message but still opening an editor... :/ WDYT?

**Comment by romartin** [comment]
@tiagodolphine @hasys anyway not really sure if it's possible to avoid opening the stunner editor in that case... The workbench editors are common/shared, once you click on an item in the explorer (left side panel), it delegates the work to the editor for the concrete type of asset, so the editor instance is being loaded and displayed. @manstis any ideas/clues plz? Can we open an editor and close it, in case the diagram is not valid for whatever reason?
At worst we can just show a message, instead of an empty canvas, and disable the toolbar buttons.... WDYT?

**Comment by tiagodolphine** [comment]
@hasys @romartin I updated the PR now the Editor is closed in case errors when loading the diagram. 
Can you please validate the behavior is as expected.
Thanks.

**Comment by romartin** [comment]
CC @hasys please let us know if you're happy with this now. Thanks!

File: kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/test/java/org/kie/workbench/common/stunner/project/client/editor/ProjectDiagramEditorTest.java [hunk 2]
```diff
@@ -253,6 +264,24 @@ public void testLoadContent() {
...
+    public void testLoadContentError() {
+        ArgumentCaptor<ServiceCallback> callbackArgumentCaptor = forClass(ServiceCallback.class);
+
+        tested.loadContent();
+
+        verify(projectDiagramServices, times(1)).getByPath(eq(path), callbackArgumentCaptor.capture());
+
+        callbackArgumentCaptor.getValue().onError(new ClientRuntimeError(new DefinitionNotFoundException()));
+
+        verify(placeManager, times( 1)).forceClosePlace(any(PathPlaceRequest.class));
```
Comment (jomarko) [inline kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/test/java/org/kie/workbench/common/stunner/project/client/editor/ProjectDiagramEditorTest.java]: Formatting, the space in `times( 1)`

File: kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/test/java/org/kie/workbench/common/stunner/project/client/editor/ProjectDiagramEditorTest.java [hunk 3]
```diff
@@ -253,6 +264,24 @@ public void testLoadContent() {
...
+
+        ArgumentCaptor<Consumer> consumerArgumentCaptor = forClass(Consumer.class);
+        verify(diagramClientErrorHandler, times(1)).handleError(any(ClientRuntimeError.class), consumerArgumentCaptor.capture());
+        consumerArgumentCaptor.getValue().accept("error message");
+        verify(errorPopupPresenter, times( 1)).showMessage(anyString());
```
Comment (jomarko) [inline kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/test/java/org/kie/workbench/common/stunner/project/client/editor/ProjectDiagramEditorTest.java]: * Formatting, the space in times( 1)
* Also I think the assert should looks like: `verify(errorPopupPresenter, times(1)).showMessage("error message");`

**Comment by hasys** [comment]
Full downstream build is broken by droolsjbpm-integration repository, working on it.

**Comment by hasys** [comment]
Jenkins execute full downstream build

**Comment by hasys** [comment]
Next know issue:
`12:00:41 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:1.4.1:enforce (ban-duplicated-classes) on project kie-wb-common-contributors-backend: Some Enforcer rules have failed. Look above for specific messages explaining why the rule failed. -> [Help 1]`

Waiting for a fix, sorry.

File: kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/test/java/org/kie/workbench/common/stunner/project/client/editor/ProjectDiagramEditorTest.java (hunk 2 above)
Comment (tiagodolphine) [reply kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/test/java/org/kie/workbench/common/stunner/project/client/editor/ProjectDiagramEditorTest.java]: @jomarko Done ! thanks

File: kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/test/java/org/kie/workbench/common/stunner/project/client/editor/ProjectDiagramEditorTest.java (hunk 3 above)
Comment (tiagodolphine) [reply kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/test/java/org/kie/workbench/common/stunner/project/client/editor/ProjectDiagramEditorTest.java]: @jomarko Done, updated the `anyString()` to the exptected string `"error message"`  ! thanks

**Comment by tiagodolphine** [comment]
@hasys I have just updated with some checkstyle issues that were solved, and building it locally all good now... let's wait for the CI tests.
Thanks !

**Comment by manstis** [comment]
@tiagodolphine Hi.. if a ```@WorkbenchEditor``` is opened for a non-existent file (e.g. a file was opened by User A in the Authoring Perspective and then the perspective closed the "open" file path is saved in the workbench "state". If User B then deletes the same file and then User A navigates to the Authoring Perspective then an attempt to open the "last open" file is made - and fails).

```KieEditor``` catches this, disables the menus and shows a "file not found" type widget in the ```@WorkbenchEditor``` panel. See [here](https://github.com/kiegroup/kie-wb-common/blob/master/kie-wb-common-widgets/kie-wb-metadata-widget/src/main/java/org/kie/workbench/common/widgets/metadata/client/KieEditor.java#L181) for some code that may give you some pointers. I assume you extend ```KieEditor``` so you could thrown a sub-class of ```NoSuchFileException``` if the diagram cannot be opened and everything should work automagically.. although you might need some refinement to show a new "Diagram cannot be opened" type message).

**Comment by tiagodolphine** [comment]
Jenkins execute full downstream build

**Comment by tiagodolphine** [comment]
Hi @manstis great to know that, do you think this is a better choice than the one I've implemented using the `placeManager.forceClosePlace(editorId)` in case of loading diagram error?
Thanks.
c/c @romartin

**Comment by manstis** [comment]
@tiagobento That's entirely up to you I guess.. I did not see you'd solved the "empty editor" issue some other way; just responding to where I saw my name mentioned! It might depend on whether you plan on showing _why_ a process could not be opened and allow the User to _fix_ the issues in which case the ```KieEditor``` approach (and a custom screen showing problems etc) might be helpful.

**Comment by hasys** [comment]
@tiagodolphine, works like a charm, thank you! :) It is already approved, but one more +1 :)

**Comment by romartin** [comment]
Jenkins please retest this

**Comment by romartin** [comment]
Jenkins please retest this

**Comment by romartin** [comment]
Jenkins please retest this

**Comment by hasys** [comment]
Jenkins retest please

**Pull request merged**